### PR TITLE
Search module

### PR DESCRIPTION
**Pull request by adrien2p**
**What**
Provide a module that allows to search data from a single place. It will allow to order, filter and retrieve data based on some configuration. It will act as the default search engine that medusa comes with and will circumvent the filtering issues that can come when having bidirectional filtering across our modules.

Also include
- module loading updates
- modules update
- orchestration fixes
- joiner fixes
- event normalization on product/inventory modules

**NOTE**
If someone relies on the product events while using the module, the event shape will be different. Same goes for the inventory

### CODE REVIEW

File: packages/medusa/src/services/product-variant-inventory.ts [hunk 1]
```diff
@@ -301,16 +317,33 @@ class ProductVariantInventoryService extends TransactionBaseService {
...
-          id: data.map((d) => d.variantId),
-        },
-        {
-          select: ["id"],
-        }
+    let variants
+    if (
+      this.featureFlagRouter_.isFeatureEnabled(
+        IsolateProductDomainFeatureFlag.key
+      )
```
Comment (pKorsholm) [inline packages/medusa/src/services/product-variant-inventory.ts:324]: nit: should be v2 FF when you merge develop

File: packages/search/integration-tests/utils/database.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
+import { TSMigrationGenerator } from "@mikro-orm/migrations"
+import { MikroORM, SqlEntityManager } from "@mikro-orm/postgresql"
+import * as ProductModels from "@models"
```
Comment (pKorsholm) [inline packages/search/integration-tests/utils/database.ts:3]: question: Shouldn't this be searchModel?

```suggestion
import * as SearchModels from "@models"
```

File: packages/search/integration-tests/utils/database.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+    return this.orm
+  },
+
+  async setupDatabase() {
+    // Initializing the ORM
+    this.orm = await MikroORM.init({
+      type: "postgresql",
+      clientUrl: DB_URL,
+      entities: Object.values(ProductModels),
```
Comment (pKorsholm) [inline packages/search/integration-tests/utils/database.ts:58]:
```suggestion
      entities: Object.values(SearchModels),
```

File: packages/search/src/scripts/migration-down.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+import { LoaderOptions, Logger } from "@medusajs/types"
+
+import { DALUtils, ModulesSdkUtils } from "@medusajs/utils"
+import { EntitySchema } from "@mikro-orm/core"
+import { SearchModuleOptions } from "../types"
+
+/**
+ * This script is only valid for mikro orm managers. If a user provide a custom manager
+ * he is in charge of reverting the migrations.
```
Comment (pKorsholm) [inline packages/search/src/scripts/migration-down.ts:11]: suggestion: Add this as a logger statement when running the script

File: packages/search/src/scripts/migration-up.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
+import { LoaderOptions, Logger } from "@medusajs/types"
+import { DALUtils, ModulesSdkUtils } from "@medusajs/utils"
+import { EntitySchema } from "@mikro-orm/core"
+import * as SearchModels from "@models"
+import { SearchModuleOptions } from "../types"
+
+/**
+ * This script is only valid for mikro orm managers. If a user provide a custom manager
+ * he is in charge of running the migrations.
```
Comment (pKorsholm) [inline packages/search/src/scripts/migration-up.ts:9]: same suggestion as above

File: packages/search/src/utils/build-config.ts [hunk 6]
```diff
@@ -0,0 +1,736 @@
...
+  if (!alias) {
+    throw new Error(
+      `CatalogModule error, the module ${relatedModule?.serviceName} has a schema but does not have any alias for the entity ${entityName}. Please add an alias to the module configuration and the entity it correspond to in the args under the entity property.`
+    )
+  }
+
+  return { relatedModule, alias }
+}
+
+// TODO: rename util
```
Comment (pKorsholm) [inline packages/search/src/utils/build-config.ts:108]: question: shoudl we handle this todo

**Review by pKorsholm — COMMENTED** [review]
Lgtm! 

I'm not 100p on everything but it generally looks good, I think merging and testing wouldn't be too bad

 there's a couple of todo's still in the code which we should either fix or remove

**Comment by adrien2p** [comment]
Find places that depends on inventory events and product events to align with the event normalization

**Pull request closed without merging**